src/main/python/deep_qa/layers/noisy_or.py

```diff
@@ -0,0 +1,88 @@
+from keras.layers import Layer
+from keras.initializations import identity
+from keras.constraints import Constraint, NonNeg
+from keras.regularizers import L1L2Regularizer
+from overrides import overrides
+
+class BetweenZeroAndOne(Constraint):
+    """Constrains the weights to be between zero and one
+    """
+    def __call__(self, p):
+        # handle negatives
+        p *= K.cast(p >= 0., K.floatx())
+        leaky_zeros_mask = K.epsilon() * K.cast(K.equal(p, K.zeros_like(p)), 'float32')
+        p = p + leaky_zeros_mask
+        # handle greater than 1
+        p *= K.cast(p <= 1., K.floatx())
+        leaky_ones_mask = 1.0 * K.cast(K.equal(p, K.zeros_like(p)), K.floatx())
+        p += leaky_ones_mask
+        return p
+
+
+class NoisyOr(Layer):
+
+    def __init__(self, axis=-1, param_init='uniform', prefix="noisy_or", **kwargs):
+        self.axis = axis
+        self.param_init = param_init
+        self.supports_masking = True
+        self.prefix = prefix
+        super(NoisyOr, self).__init__(**kwargs)
+
+    def build(self, input_shape):
+        # Create a trainable weight variable for the noise parameter
+        self.noise_parameter = self.add_weight(shape=(1,),
+                                               name=self.prefix + '_q',
+                                               initializer=self.param_init,
+                                               regularizer=L1L2Regularizer(l2=0.001),
+                                               # constraint=BetweenZeroAndOne(),
+                                 trainable=True)
+        super(NoisyOr, self).build(input_shape)
+
+    def get_output_shape_for(self, input_shape):
+        if self.axis == -1:
+            return input_shape[0:-1]
+        return input_shape[:self.axis - 1] + input_shape[self.axis:]
+
+    @overrides
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        # We do not need a mask beyond this layer.
+        # (probably right)
+        if mask is not None:
+            return K.any(mask, axis=-1)
+        return None
+
+    def call(self, x, mask=None):
+        """
+        Calculate the noisy-or probability across a given tensor axis based on
+        noisy-or equation:
+            p(x) = 1 - Prod_(i=1:N)[1 - q * p(x|y_n)]
+
+        Parameters
+        ----------
+        inputs: Tensor
+
+        mask: Tensor or None, optional (default=None)
+            Tensor of shape (batch size, max number of options) representing
+            which options are padding and thus have a 0 in the associated
+            mask position.
+
+        Returns
+        -------
+        output: Tensor
+            The combined probability of the even happening, given the parent events
+            Output dims are one less than input dims
+        """
+        # dim: (b, ..., num_probs)
+        noisy_probs = self.noise_parameter * x
+
+        # dim: (b, ..., num_probs)
+        noisy_probs = K.ones_like(x) - noisy_probs
```

Comment:
Does doing `1.0 - noisy_probs` or `noisy_probs * -1 + 1` work?  If it did, you could avoid a tensor allocation.  Not sure if it does or not (and I added the second one because of how operator overloading works; it might work if the first doesn't).